### PR TITLE
[KMSv2] promote KMSv2 and KMSv2KDF to GA

### LINKED ISSUE #120309: [KMSv2] update code to mark API as GA in v1.29
I think all of the changes will be focused around allowing `v2` as an acceptable value in addition to `v2beta1` (with associated unit/integration tests):

https://github.com/kubernetes/kubernetes/blob/76b06ee239a893dece6eae861090f06df994a878/staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go#L55-L56

/sig auth
/triage accepted
/assign ritazh
/milestone v1.29

### PR DESCRIPTION
**Pull request by ritazh**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Promote KMSv2 and KMSv2KDF to GA

#### Which issue(s) this PR fixes:

Fixes #120309

#### Does this PR introduce a user-facing change?

```release-note
The KMSv2 feature with feature gates `KMSv2` and `KMSv2KDF` are promoted to GA.  The `KMSv1` feature gate is now disabled by default.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
[KEP] https://github.com/kubernetes/enhancements/tree/master/keps/sig-auth/3299-kms-v2-improvements
```

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go [hunk 1]
```diff
@@ -51,7 +51,7 @@ func init() {
 
 const (
 	// KMSAPIVersion is the version of the KMS API.
-	KMSAPIVersion = "v2beta1"
+	KMSAPIVersion = "v2"
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/envelope.go:54]: We have to support both (likely forever because there is no good reason to break clients).

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/testing/v2/kms_plugin_mock.go [hunk 2]
```diff
@@ -43,7 +43,7 @@ const (
 	unixProtocol = "unix"
 
 	// Current version for the protocol interface definition.
-	kmsapiVersion = "v2beta1"
+	kmsapiVersion = "v2"
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/testing/v2/kms_plugin_mock.go:46]: We need a test that shows that both values work correctly.

File: staging/src/k8s.io/kms/internal/plugins/_mock/pkcs11/pkcs11.go [hunk 3]
```diff
@@ -111,7 +111,7 @@ func (s *pkcs11RemoteService) Decrypt(ctx context.Context, uid string, req *serv
 
 func (s *pkcs11RemoteService) Status(ctx context.Context) (*service.StatusResponse, error) {
 	return &service.StatusResponse{
-		Version: "v2beta1",
+		Version: "v2",
```
Comment (enj) [inline staging/src/k8s.io/kms/internal/plugins/_mock/pkcs11/pkcs11.go:114]: This is okay to change since this is our "example" plugin we want people to copy

File: staging/src/k8s.io/apiserver/pkg/features/kube_features.go [hunk 4]
```diff
@@ -290,9 +292,9 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 
 	KMSv1: {Default: true, PreRelease: featuregate.Deprecated},
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/features/kube_features.go:293]: Flip this to disabled by default.

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go [hunk 5]
```diff
@@ -111,7 +111,7 @@ func (t *testKMSv2EnvelopeService) Status(ctx context.Context) (*kmsservice.Stat
 	if t.err != nil {
 		return nil, t.err
 	}
-	return &kmsservice.StatusResponse{Healthz: "ok", KeyID: t.keyID, Version: "v2beta1"}, nil
+	return &kmsservice.StatusResponse{Healthz: "ok", KeyID: t.keyID, Version: "v2"}, nil
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go:114]: Undo this change.

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go [hunk 6]
```diff
@@ -1871,36 +1822,6 @@ func Test_kmsv2PluginProbe_rotateDEKOnKeyIDChange(t *testing.T) {
...
-			state:       envelopekmsv2.State{},
-			statusKeyID: "1",
-			wantState: envelopekmsv2.State{
-				EncryptedObject:     kmstypes.EncryptedObject{KeyID: "1"},
-				ExpirationTimestamp: now.Add(3 * time.Minute),
-			},
-			wantEncryptCalls: 1,
-			wantLogs: []string{
-				`"encrypting content using envelope service" uid="panda"`,
-				fmt.Sprintf(`"successfully rotated DEK" uid="panda" useSeed=false newKeyIDHash="sha256:6b86b273ff34fce19d6b804eff5a3f5747ada4eaa22f1d49c01e52ddb7875b4b" oldKeyIDHash="" expirationTimestamp="%s"`,
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go:1886]: You need to come up with a way to not delete these tests.  We cannot lose coverage for the `useSeed=false` path.

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go [hunk 7]
```diff
@@ -1775,7 +1726,7 @@ func TestIsKMSv2ProviderHealthyError(t *testing.T) {
 			desc: "missing keyID",
 			statusResponse: &kmsservice.StatusResponse{
 				Healthz: "ok",
-				Version: "v2beta1",
+				Version: "v2",
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go:1729]: Undo (applies multiple times)

File: test/integration/controlplane/transformation/kmsv2_transformation_test.go [hunk 8]
```diff
@@ -172,12 +171,6 @@ func (r envelopekmsv2) plainTextPayload(secretETCDPath string) ([]byte, error) {
 // 5. kmstypes.EncryptedObject structure should be serialized and deposited in ETCD
 func TestKMSv2Provider(t *testing.T) {
 	defaultUseSeed := utilfeature.DefaultFeatureGate.Enabled(features.KMSv2KDF)
-
-	t.Run("regular gcm", func(t *testing.T) {
-		defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.KMSv2KDF, false)()
-		testKMSv2Provider(t, !defaultUseSeed)
```
Comment (enj) [inline test/integration/controlplane/transformation/kmsv2_transformation_test.go:178]: Hmm I need to think through if this is okay to delete.

File: staging/src/k8s.io/apiserver/pkg/features/kube_features.go [hunk 9]
```diff
@@ -290,9 +292,9 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
...
 
-	KMSv2: {Default: true, PreRelease: featuregate.Beta},
+	KMSv2: {Default: true, PreRelease: featuregate.GA, LockToDefault: true}, // remove in 1.31
 
-	KMSv2KDF: {Default: true, PreRelease: featuregate.Beta}, // lock to true in 1.29 once KMSv2 is GA, remove in 1.31
+	KMSv2KDF: {Default: true, PreRelease: featuregate.GA, LockToDefault: true}, // remove in 1.31
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/features/kube_features.go:297]: @liggitt @deads2k @mikedanese how do we feel about `KMSv2KDF` being locked to `true`?

**Review by enj — CHANGES_REQUESTED** [review]

**Comment by MaryamTavakkoli** [comment]
Hello!
Bug triage for the 1.29 release cycle is here!
The code freeze will start ([01:00 UTC Wednesday 1st November 2023 / 18:00 PDT Tuesday 31st October 2023](https://everytimezone.com/s/3b849751)), which is about 1 week from now. And while there is still plenty of time, we want to make sure that every PR has a chance to be merged on time.

As the PR is tagged for 1.29, is it still planned for that release?

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go [hunk 10]
```diff
@@ -96,6 +96,7 @@ const (
 	kmsPluginHealthzPositiveTTL = 20 * time.Second
 	kmsAPIVersionV1             = "v1"
 	kmsAPIVersionV2             = "v2"
+	kmsAPIVersionV2beta1        = "v2beta1"
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:99]: Undo this change, the config file has only ever supported the major version.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/grpc_service.go
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/grpc_service.go:1]: Undo changes to this file, it should not care about version.

File: staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/grpc_service_unix_test.go
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/storage/value/encrypt/envelope/kmsv2/grpc_service_unix_test.go:1]: Undo changes to this file.

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go [hunk 11]
```diff
@@ -720,7 +721,8 @@ func kmsPrefixTransformer(ctx context.Context, config *apiserverconfig.KMSConfig
 			// for v1 we will do encrypt and decrypt for health check. Since these are serial operations, we will double the timeout.
 			kmsTimeoutSum: 2 * config.Timeout.Duration,
 		}, nil
-
+	case kmsAPIVersionV2beta1:
+		fallthrough
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:725]: Undo.

File: test/integration/controlplane/transformation/kms_transformation_test.go [hunk 12]
```diff
@@ -596,8 +601,11 @@ resources:
 
 	t.Run("encrypt all resources", func(t *testing.T) {
 		_ = mock.NewBase64Plugin(t, "@encrypt-all-kms-provider.sock")
-		defer featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, "AllAlpha", true)()
-		defer featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, "AllBeta", true)()
+		// defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, "AllAlpha", true)()
+		// defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, "AllBeta", true)()
+		defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.KMSv1, true)()
+		defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.KMSv2, true)()
+
```
Comment (enj) [inline test/integration/controlplane/transformation/kms_transformation_test.go:608]: Undo this change.

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go [hunk 13]
```diff
@@ -188,6 +188,7 @@ func TestLegacyConfig(t *testing.T) {
 
 func TestEncryptionProviderConfigCorrect(t *testing.T) {
 	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.KMSv2, true)()
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.KMSv1, true)()
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go:191]: Confirm that we have a unit test that checks that KMS v1 cannot be used when disabled.

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go [hunk 14]
```diff
@@ -615,6 +568,16 @@ func TestKMSMaxTimeout(t *testing.T) {
...
+									APIVersion: "v2beta1",
+									Timeout: &metav1.Duration{
+										Duration: 5 * time.Second,
+									},
+									Endpoint: "unix:///tmp/anothertestprovider.sock",
+								},
+							},
+							{
+								KMS: &apiserverconfig.KMSConfiguration{
+									Name:       "new-kms2",
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go:580]: Undo.

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go [hunk 15]
```diff
@@ -454,8 +455,8 @@ func (h *kmsv2PluginProbe) isKMSv2ProviderHealthyAndMaybeRotateDEK(ctx context.C
 	if response.Healthz != "ok" {
 		errs = append(errs, fmt.Errorf("got unexpected healthz status: %s", response.Healthz))
 	}
-	if response.Version != envelopekmsv2.KMSAPIVersion {
-		errs = append(errs, fmt.Errorf("expected KMSv2 API version %s, got %s", envelopekmsv2.KMSAPIVersion, response.Version))
+	if response.Version != envelopekmsv2.KMSAPIVersionv2 && response.Version != envelopekmsv2.KMSAPIVersionv2beta1 {
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:458]: You also need to make sure that the version does not change after the first status response.

File: test/integration/controlplane/transformation/kmsv2_transformation_test.go [hunk 16]
```diff
@@ -768,7 +717,7 @@ func assertPodDEKSources(ctx context.Context, t *testing.T, config storagebacken
 
 func TestKMSv2Healthz(t *testing.T) {
 	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.KMSv2, true)()
-	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.KMSv2KDF, randomBool())()
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.KMSv2KDF, true)()
```
Comment (enj) [inline test/integration/controlplane/transformation/kmsv2_transformation_test.go:720]: Lets replace the wiring for this to look at an atomic bool that we can swap in tests (but cannot be set by an end user).  Then we can avoid having to make any large changes to our tests.

**Review by enj — CHANGES_REQUESTED** [review]

**Comment by enj** [comment]
/release-note-edit The KMSv2 feature with feature gates `KMSv2` and `KMSv2KDF` are promoted to GA.  The KMSv1 feature gate is now disabled by default.

File: staging/src/k8s.io/apiserver/pkg/features/kube_features.go (hunk 9 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/features/kube_features.go:297]: sgtm, that was the plan with v2 graduation

**Comment by enj** [comment]
/release-note-edit ```release-note
The KMSv2 feature with feature gates `KMSv2` and `KMSv2KDF` are promoted to GA.  The `KMSv1` feature gate is now disabled by default.
```

File: test/integration/controlplane/transformation/kms_transformation_test.go (hunk 12 above)
Comment (ritazh) [reply test/integration/controlplane/transformation/kms_transformation_test.go:608]: We still need `defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.KMSv1, true)()` since some of these tests require kmsv1 to be enabled.
Note this is kms_transformation_test.go not kmsv2_transformation_test.go

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go (hunk 15 above)
Comment (ritazh) [reply staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:458]: added a check and a test

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go (hunk 13 above)
Comment (ritazh) [reply staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go:191]: added

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go [hunk 17]
```diff
@@ -107,6 +107,19 @@ const (
 
 var codecs serializer.CodecFactory
 
+var kdfDisabled atomic.Bool
+
+func SetKDFForTests(b bool) func() {
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go:112]: Add comments describing why we have these functions.

File: staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go [hunk 18]
```diff
@@ -2065,7 +2072,7 @@ func Test_kmsv2PluginProbe_rotateDEKOnKeyIDChange(t *testing.T) {
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.KMSv2KDF, tt.useSeed)()
+			SetKDFForTests(tt.useSeed)
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config_test.go:2075]:
```suggestion
			defer SetKDFForTests(tt.useSeed)()
```

File: test/integration/controlplane/transformation/kmsv2_transformation_test.go [hunk 19]
```diff
@@ -1317,12 +1269,11 @@ func randomBool() bool { return utilrand.Int()%2 == 1 }
...
 func TestKMSv2ProviderLegacyData(t *testing.T) {
 	t.Run("regular gcm", func(t *testing.T) {
-		defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.KMSv2KDF, false)()
+		defer encryptionconfig.SetKDFForTests(false)()
 		testKMSv2ProviderLegacyData(t)
 	})
-
 	t.Run("extended nonce gcm", func(t *testing.T) {
-		defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.KMSv2KDF, true)()
+		// default KMSv2KDF feature is enabled
```
Comment (enj) [inline test/integration/controlplane/transformation/kmsv2_transformation_test.go:1276]: This isn't testing the default, so it should explicitly set the value it wants.

```suggestion
		defer encryptionconfig.SetKDFForTests(true)()
```

File: test/integration/controlplane/transformation/kmsv2_transformation_test.go [hunk 20]
```diff
@@ -1231,8 +1184,7 @@ func BenchmarkKMSv2REST(b *testing.B) {
 	klog.SetOutput(io.Discard)
 	klog.LogToStderr(false)
 
-	defer featuregatetesting.SetFeatureGateDuringTest(b, utilfeature.DefaultFeatureGate, features.KMSv2, true)()
-	defer featuregatetesting.SetFeatureGateDuringTest(b, utilfeature.DefaultFeatureGate, features.KMSv2KDF, false)()
+	defer featuregatetesting.SetFeatureGateDuringTest(b, utilfeature.DefaultFeatureGate, features.KMSv2KDF, true)()
```
Comment (enj) [inline test/integration/controlplane/transformation/kmsv2_transformation_test.go:1187]:
```suggestion
	defer encryptionconfig.SetKDFForTests(true)()
```

File: test/integration/controlplane/transformation/kmsv2_transformation_test.go [hunk 21]
```diff
@@ -174,12 +187,11 @@ func TestKMSv2Provider(t *testing.T) {
 	defaultUseSeed := utilfeature.DefaultFeatureGate.Enabled(features.KMSv2KDF)
```
Comment (enj) [inline test/integration/controlplane/transformation/kmsv2_transformation_test.go:187]:
```suggestion
	defaultUseSeed := encryptionconfig.GetKDF()
```

File: test/integration/controlplane/transformation/kmsv2_transformation_test.go [hunk 22]
```diff
@@ -174,12 +187,11 @@ func TestKMSv2Provider(t *testing.T) {
...
 
 	t.Run("regular gcm", func(t *testing.T) {
-		defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.KMSv2KDF, false)()
+		defer encryptionconfig.SetKDFForTests(!defaultUseSeed)()
```
Comment (enj) [inline test/integration/controlplane/transformation/kmsv2_transformation_test.go:190]:
```suggestion
		defer encryptionconfig.SetKDFForTests(false)()
```

File: test/integration/controlplane/transformation/kmsv2_transformation_test.go [hunk 23]
```diff
@@ -174,12 +187,11 @@ func TestKMSv2Provider(t *testing.T) {
...
 		testKMSv2Provider(t, !defaultUseSeed)
 	})
-
 	t.Run("extended nonce gcm", func(t *testing.T) {
-		defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.KMSv2KDF, true)()
+		defer encryptionconfig.SetKDFForTests(defaultUseSeed)()
```
Comment (enj) [inline test/integration/controlplane/transformation/kmsv2_transformation_test.go:194]:
```suggestion
		defer encryptionconfig.SetKDFForTests(true)()
```

File: test/integration/controlplane/transformation/kmsv2_transformation_test.go [hunk 24]
```diff
@@ -163,6 +163,19 @@ func (r envelopekmsv2) plainTextPayload(secretETCDPath string) ([]byte, error) {
 	return plainSecret, nil
 }
 
+// TestDefaultValues tests default flag values without updating any of the feature flags
+func TestDefaultValues(t *testing.T) {
```
Comment (enj) [inline test/integration/controlplane/transformation/kmsv2_transformation_test.go:167]: Run the server with KMS v2 configured (without setting any feature gate or calling `SetKDFForTests`) and assert that the data stored in etcd is using KDF.

File: test/integration/controlplane/transformation/kmsv2_transformation_test.go [hunk 25]
```diff
@@ -927,8 +929,7 @@ resources:
 	pluginMock := kmsv2mock.NewBase64Plugin(t, "@kms-provider.sock")
 	storageConfig := framework.SharedEtcd()
 
-	// When feature flag is enabled, loading a encryptionConfig with KMSv1 and v2 should work
-	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.KMSv2, true)()
+	// KMSv2is enabled by default. Loading a encryptionConfig with KMSv2 should work
```
Comment (enj) [inline test/integration/controlplane/transformation/kmsv2_transformation_test.go:932]:
```suggestion
	// KMSv2 is enabled by default. Loading a encryptionConfig with KMSv2 should work
```

File: staging/src/k8s.io/apiserver/pkg/features/kube_features.go [hunk 26]
```diff
@@ -288,11 +290,11 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
...
 	EfficientWatchResumption: {Default: true, PreRelease: featuregate.GA, LockToDefault: true},
 
-	KMSv1: {Default: true, PreRelease: featuregate.Deprecated},
+	KMSv1: {Default: false, PreRelease: featuregate.Deprecated},
 
-	KMSv2: {Default: true, PreRelease: featuregate.Beta},
+	KMSv2: {Default: true, PreRelease: featuregate.GA, LockToDefault: true}, // remove in 1.31
 
-	KMSv2KDF: {Default: true, PreRelease: featuregate.Beta}, // lock to true in 1.29 once KMSv2 is GA, remove in 1.31
+	KMSv2KDF: {Default: true, PreRelease: featuregate.GA, LockToDefault: true}, // remove in 1.31
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/features/kube_features.go:297]: There are 4 uses of this feature gate in tests that need to replaced with calls to Get/Set.  i.e. at the end of this PR, this feature gate should not be used at all.

**Comment by enj** [comment]
/release-note-edit ```release-note The KMSv2 feature with feature gates `KMSv2` and `KMSv2KDF` are promoted to GA.  The `KMSv1` feature gate is now disabled by default.```

**Comment by enj** [comment]
/release-note-edit
```release-note
The KMSv2 feature with feature gates `KMSv2` and `KMSv2KDF` are promoted to GA.  The `KMSv1` feature gate is now disabled by default.
```

**Review by enj — CHANGES_REQUESTED** [review]
This is close.

**Comment by enj** [comment]
@aramase the kms test failures look strange, any ideas?

/test pull-kubernetes-e2e-kind-kms

File: staging/src/k8s.io/apiserver/pkg/features/kube_features.go [hunk 27]
```diff
@@ -288,11 +290,11 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 
 	EfficientWatchResumption: {Default: true, PreRelease: featuregate.GA, LockToDefault: true},
 
-	KMSv1: {Default: true, PreRelease: featuregate.Deprecated},
+	KMSv1: {Default: false, PreRelease: featuregate.Deprecated},
 
-	KMSv2: {Default: true, PreRelease: featuregate.Beta},
+	KMSv2: {Default: true, PreRelease: featuregate.GA, LockToDefault: true}, // remove in 1.31
```
Comment (enj) [inline staging/src/k8s.io/apiserver/pkg/features/kube_features.go:295]: There are ~10 usages of this feature gate that need to be removed.

**Comment by enj** [comment]
> @aramase the kms test failures look strange, any ideas?
> 
> /test pull-kubernetes-e2e-kind-kms

#121543

File: test/integration/controlplane/transformation/kmsv2_transformation_test.go [hunk 28]
```diff
@@ -163,6 +162,29 @@ func (r envelopekmsv2) plainTextPayload(secretETCDPath string) ([]byte, error) {
...
+		t.Fatalf("without updating the feature flags, default value of KMSv1 should be disabled.")
+	}
+	// since encryptionconfig.GetKDF() is true by default, following test should verify if
+	// object.EncryptedDEKSourceType == kmstypes.EncryptedDEKSourceType_HKDF_SHA256_XNONCE_AES_GCM_SEED
+	testKMSv2ProviderDEKSourceReuse(t,
+		func(_ int, _ uint64, etcdKey string, obj kmstypes.EncryptedObject) {
+			if obj.KeyID != "1" {
+				t.Errorf("key %s: want key ID %s, got %s", etcdKey, "1", obj.KeyID)
+			}
+		},
```
Comment (enj) [inline test/integration/controlplane/transformation/kmsv2_transformation_test.go:184]: Check `obj.EncryptedDEKSourceType == kmstypes.EncryptedDEKSourceType_HKDF_SHA256_XNONCE_AES_GCM_SEED` manually in the closure.  I know it is a duplicate check but it makes it much easier to reason about by explicitly checking it here.  A future change to `testKMSv2ProviderDEKSourceReuse` could invalidate this test without us realizing it.

**Review by enj — CHANGES_REQUESTED** [review]
Feel free to squash after this.

**Comment by liggitt** [comment]
/approve
For feature gate
